keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -9,105 +9,201 @@
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
+message SignJWTRequest {
+  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
   bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message SignJWTResponse {
+  // The header for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  bytes header = 1;
+
+  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  bytes signature = 2;
 }
 
+message FetchKeysRequest {}
+
+message FetchKeysResponse {
+  repeated Key keys = 1;
+
+  // The timestamp when this data was pulled from the authoritative source of
+  // truth for verification keys.
+  google.protobuf.Timestamp data_timestamp = 2;
 
-message PublicKey {
-  // public_key is a PEM encoded public key
-  bytes public_key = 1;
-  // certificate is a concatenated list of PEM encoded x509 certificates
-  bytes certificates = 2;
-  // key_id is the key's ID
-  string key_id = 3;
-  // algorithm states the algorithm the key uses
-  string algorithm = 4;
+  // The timestamp for when it's appropriate to refetch verification keys to 
+  // pick changes if any.
+  google.protobuf.Timestamp suggested_next_fetch_at = 3;
 }
 
-message ListPublicKeysRequest {}
-message ListPublicKeysResponse {
-  // key_id is the key's ID
-  string active_key_id = 1;
-  // public_keys is a list of public verifying keys
-  repeated PublicKey public_keys = 2;
+message Key {
+  // A unique identifier for this key.
+  string key_id = 1;
+
+  // The public key, PKIX-serialized.
+  bytes key = 2;
 }
 ```
 
-### Implementation Details/Notes/Constraints
+### Test Plan
 
-The API server flag `--service-account-key-file` can be specified multiple times for legacy SA tokens and projected tokens. Validation keys from this flag will be merged with the response of `ListPublicKeys()`. A new flag `--key-service-url` will be added to the API server specifying a unix socket where the key service will be accessible.
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
 
-### Risks and Mitigations
+##### Prerequisite testing updates
+
+<!-- TODO -->
+
+##### Unit tests
+
+Will need coverage for:
+
+- pkg/kubeapiserver/options/authentication.go
+- the signer plugin 
+- the key cache plugin
+
+<!-- TODO
+Add info on concerned test in below form as read from https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+- `<package>`: `<date>` - `<test coverage>`
+ -->
+
```

Comment:
add an integration test section describing wiring up a real working kube-apiserver configured to run with this signer (KMS KEP might have a good testing section you can mimic)

Reply:
Added some initial bits. Can you point me to the KEP that you are talking about?

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements#integration-tests